hell/adventcalendar/2025/3/index.md

```diff
@@ -1,35 +1,123 @@
 
-## Heading
+[https://schepp.dev/](Schepp) covered [the API in last year's post](/adventcalendar/2024/28/) and discussed the `eagerness` value which allowed you to, for example, hover over a link to speculatively prerender it with the simple addition of this rule to your HTML:
 
-Some text.
+```html
+<script type="speculationrules">
+{
+  "prerender": [{
+    "where": { "href_matches": "/*" },
+    "eagerness": "moderate"
+  }]
+}
+</script>
+```
+
+This hover technique is common functionality offered by libraries, plugins and frameworks but now this is baked right into the browser. Chromium-based browsers only for now but it is being worked on by Safari and Firefox, and as a progressive enhancement nothing is broken on non-supporting browsers — they just don't benefit from the performance improvement.
+
+Using mouse hover as a signal works well on desktop, but what about mobile where "hover" is not a really a thing? At least in browser terms — I'm sure many of us have hovered over a link with out fingers but as that is not detected by the browser as an event, we can't act upon it.
+
+Well solving that is just one of many improvements the API has seen in the last year that we'll cover in this post.
+
+## Improved mobile viewport heuristics
+
+Since hover does not exist, we have updated the `eagerness: moderate` on mobile to instead look at the following heuristics:
+
+- Anchor links within 30% vertical distance from the previous pointer down.
+- Anchor links at least 0.5x as big as the largest anchor in the viewport.
+- The browser waits 500 milliseconds after the user stopped scrolling.
+
+The aim of this is to try to avoid, as much as possible, over-speculating links users are less likely to click on. For example, because they are still scrolling, or they are very small links (terms and conditions and the like!).
+
+You can see this in action in the following video where the [Web Almanac](https://almanac.httparchiv.org) prerenders same-origin links with `moderate` eagerness:
+
+<figure>
+  <video controls playsinline muted loop>
+      <source src="/images/advent2025/speculation-rules-mobile-viewport-demo.mp4" width=2420 height=1320>
+  </video>
+  <figcaption><p>An example of scrolling down a mobile page triggering speculations.</p></figcaption>
+</figure>
+<script>
+  const mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)');
+  if (!mediaQuery.matches) {
+      document.querySelector('video').autoplay = true;
+  }
+</script>
+
+In the video you can see as you scroll down the page, links are being successfully prerendered ready for the user to browse to the page.
+
+<p class="highlight"><strong>Note:</strong> Speculating has a cost. To both your users, and for sites with potential increased traffic and and resulting infrastructure usage. Always weigh those costs with the benefit to the user.</p>
```

Comment:
Now I'm wondering how dangerous it is. Can we take the first code snippet from your post, add it to our HTML boilerplate, and use it on every site?
How do we determine if it's over-speculating, and how do we deal with it?